src/perturbed/abstract_perturbed.jl

```diff
@@ -70,18 +69,64 @@ Turn random perturbations of `θ` into a distribution on polytope vertices.
 Keyword arguments are passed to the underlying linear maximizer.
 """
 function compute_probability_distribution(
-    perturbed::AbstractPerturbed, θ::AbstractArray; kwargs...
+    perturbed::AbstractPerturbed,
+    θ::AbstractArray;
+    autodiff_variance_reduction::Bool=false,
+    kwargs...,
 )
-    Z_samples = sample_perturbations(perturbed, θ)
-    return compute_probability_distribution(perturbed, θ, Z_samples; kwargs...)
+    η_samples = sample_perturbations(perturbed, θ)
+    return compute_probability_distribution(perturbed, η_samples; kwargs...)
 end
 
+# Forward pass
+
 """
     (perturbed::AbstractPerturbed)(θ; kwargs...)
 
 Apply `compute_probability_distribution(perturbed, θ; kwargs...)` and return the expectation.
 """
-function (perturbed::AbstractPerturbed)(θ::AbstractArray; kwargs...)
-    probadist = compute_probability_distribution(perturbed, θ; kwargs...)
+function (perturbed::AbstractPerturbed)(
+    θ::AbstractArray; autodiff_variance_reduction::Bool=false, kwargs...
+)
+    probadist = compute_probability_distribution(
+        perturbed, θ; autodiff_variance_reduction, kwargs...
+    )
     return compute_expectation(probadist)
 end
+
+# Backward pass
+
+function ChainRulesCore.rrule(
+    rc::RuleConfig,
+    ::typeof(compute_probability_distribution),
+    perturbed::AbstractPerturbed,
+    θ::AbstractArray;
+    autodiff_variance_reduction::Bool=false,
+    kwargs...,
+)
+    η_samples = sample_perturbations(perturbed, θ)
+    y_dist = compute_probability_distribution(perturbed, η_samples; kwargs...)
+
+    ∇logp_samples = [perturbation_grad_logdensity(rc, perturbed, θ, η) for η in η_samples]
+
+    M = get_nb_samples(perturbed)
+    function perturbed_oracle_dist_pullback(δy_dist)
+        weights = y_dist.weights
+        δy_weights = δy_dist.weights
+        δy_sum = sum(δy_weights)
+        δθ = sum(
+            map(1:M) do i
+                δyᵢ, ∇logpᵢ, w = δy_weights[i], ∇logp_samples[i], weights[i]
+                if autodiff_variance_reduction
+                    bᵢ = M == 1 ? 0 * δy_sum : (δy_sum - δyᵢ) / (M - 1)
```

Comment:
Do we want to average with respect to the w distribution or as we did in the former rrules for perturbed additive and multiplicative with equal weights on the atoms ?